### PR TITLE
Protect against user function modifying input slice.

### LINKED ISSUE #24: fd: guard against function modification
Gradient at the moment will fail if the function modifies the input slice. We could say the function is not allowed to, but there's no good reason to not protectively copy

### PR DESCRIPTION
**Pull request by btracey**
Fixes #24.

### CODE REVIEW

**Comment by btracey** [comment]
PTAL @vladimir-ch 

While it is a fair amount of extra copying, keep in mind that copying is basically the fastest thing you can do in Go (it's already coded with SIMD), and that the gradient is totally worthless if the user does modify the slice. If you still think it's too much, let me know.

File: fd/diff.go [hunk 1]
```diff
@@ -175,15 +174,14 @@ func Gradient(dst []float64, f func([]float64) float64, x []float64, settings *S
...
 		go func(sendChan <-chan fdrun, ansChan chan<- fdrun, quit <-chan struct{}) {
 			xcopy := make([]float64, len(x))
-			copy(xcopy, x)
 			for {
 				select {
 				case <-quit:
 					return
 				case run := <-sendChan:
+					copy(xcopy, x)
 					xcopy[run.idx] += run.pt.Loc * step
```
Comment (vladimir-ch) [inline fd/diff.go]: Note that we should be smarter about this - if the formula contains the origin, we are re-evaluating the function there len(x) times.

**Comment by vladimir-ch** [comment]
It can be quite a lot but I don't know if it's too much. Let's be safe for now, later I'll try to profile it for example with the minimal surface function we have.

LGTM

**Comment by vladimir-ch** [comment]
Output of `go tool pprof` and `time` for the minimization of `functions.NewMinimalSurface(100, 100)` with `fd.Gradient` and LBFGS. If `runtime.memmove` comes from `copy`, then its effect seems to be pretty negligible, especially in the serial case. The speedup from concurrency is amazing. Of course, with the analytical gradient the whole problem is solved in miliseconds with 147 iterations, these results are the first 30 iterations.

`Concurrent: false`:
```
108.14s of 108.22s total (99.93%)
Dropped 25 nodes (cum <= 0.54s)
      flat  flat%   sum%        cum   cum%
    71.74s 66.29% 66.29%    107.39s 99.23%  github.com/gonum/optimize/functions.(*MinimalSurface).Func
    35.65s 32.94% 99.23%     35.65s 32.94%  github.com/gonum/optimize/functions.(*MinimalSurface).at
     0.74s  0.68% 99.92%      0.74s  0.68%  runtime.memmove
     0.01s 0.0092% 99.93%    107.40s 99.24%  github.com/gonum/optimize/functions.(*MinimalSurface).Func-fm
         0     0% 99.93%    108.14s 99.93%  github.com/gonum/diff/fd.Gradient

107.79s user 0.45s system 100% cpu 1:48.21 total
```

`Concurrent: true`:
```
108.04s of 108.30s total (99.76%)
Dropped 38 nodes (cum <= 0.54s)
      flat  flat%   sum%        cum   cum%
    71.99s 66.47% 66.47%    106.81s 98.62%  github.com/gonum/optimize/functions.(*MinimalSurface).Func
    34.82s 32.15% 98.62%     34.82s 32.15%  github.com/gonum/optimize/functions.(*MinimalSurface).at
     1.22s  1.13% 99.75%      1.22s  1.13%  runtime.memmove
     0.01s 0.0092% 99.76%    108.18s 99.89%  github.com/gonum/diff/fd.Gradient.func1
         0     0% 99.76%    106.81s 98.62%  github.com/gonum/optimize/functions.(*MinimalSurface).Func-fm
         0     0% 99.76%    108.29s   100%  runtime.goexit

107.92s user 0.20s system 570% cpu 18.946 total
```

**Pull request merged**